src/view/renderer.js

```diff
@@ -494,52 +643,78 @@ export default class Renderer {
 			this._addInlineFiller( domDocument, expectedDomChildren, filler.offset );
 		}
 
-		const actions = diff( actualDomChildren, expectedDomChildren, sameNodes );
+		return {
+			actions: diff( actualDomChildren, expectedDomChildren, sameNodes ),
+			domElement,
+			actualDomChildren,
+			expectedDomChildren
+		};
+	}
 
-		let i = 0;
-		const nodesToUnbind = new Set();
+	/**
+	 * Finds DOM nodes which were replaced with the similar nodes (same tag name) in the view. All nodes are compared
+	 * within one `insert`/`delete` action group, for example:
+	 *
+	 * 		Actual DOM:		<p><b>Foo</b>Bar<i>Baz</i><b>Bax</b></p>
+	 * 		Expected DOM:	<p>Bar<b>123</b><i>Baz</i><b>456</b></p>
+	 * 		Input actions:	[ insert, insert, delete, delete, equal, insert, delete ]
+	 * 		Output actions:	[ insert, replace, delete, equal, replace ]
+	 *
+	 * @private
+	 * @param {Array} actions Actions array which is result of {@link module:utils/diff~diff} function.
+	 * @param {Array} actualDom Actual DOM children
+	 * @param {Array} expectedDom Expected DOM children.
+	 * @returns {Array} Actions array modified with `replace` actions.
+	 */
+	_findReplaceActions( actions, actualDom, expectedDom ) {
+		// If there is no both 'insert' and 'delete' actions, no need to check for replaced elements.
+		if ( actions.indexOf( 'insert' ) === -1 || actions.indexOf( 'delete' ) === -1 ) {
+			return actions;
+		}
+
+		function areSimilar( domNode1, domNode2 ) {
+			return isNode( domNode1 ) && isNode( domNode2 ) &&
+				!isText( domNode1 ) && !isText( domNode2 ) &&
+				domNode1.tagName.toLowerCase() === domNode2.tagName.toLowerCase();
+		}
+
+		function calculateReplaceActions( actual, expected ) {
+			return diff( actual, expected, areSimilar ).map( x => x === 'equal' ? 'replace' : x );
+		}
 
+		let newActions = [];
+		let skipActions = [];
+		let actualSlice = [];
+		let expectedSlice = [];
+
+		const counter = { equal: 0, insert: 0, delete: 0 };
 		for ( const action of actions ) {
 			if ( action === 'insert' ) {
-				insertAt( domElement, i, expectedDomChildren[ i ] );
-				i++;
+				skipActions.push( 'insert' );
+				expectedSlice.push( expectedDom[ counter.equal + counter.insert ] );
 			} else if ( action === 'delete' ) {
-				nodesToUnbind.add( actualDomChildren[ i ] );
-				remove( actualDomChildren[ i ] );
-			} else { // 'equal'
-				// Force updating text nodes inside elements which did not change and do not need to be re-rendered (#1125).
-				this._markDescendantTextToSync( domConverter.domToView( expectedDomChildren[ i ] ) );
-				i++;
+				skipActions.push( 'delete' );
+				actualSlice.push( actualDom[ counter.equal + counter.delete ] );
+			} else { // equal
+				if ( expectedSlice.length && actualSlice.length ) {
+					newActions = newActions.concat( calculateReplaceActions( actualSlice, expectedSlice ) );
+				} else if ( expectedSlice.length || actualSlice.length ) {
+					newActions = newActions.concat( skipActions );
```

Comment:
Is this needed? Wouldn't `calculateReplaceActions`, or rather `diff`, return `[ 'insert', 'insert', ... ]` if `actualSlice` is empty and `[ 'delete', 'delete', ... ]` if `expectedSlice` is empty?

Reply:
Especially since you don't have this below (in line 713) which is the same situation as meeting `equal`.